samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -0,0 +1,747 @@
+		.data.ptr = msg,
+		.data.len = strlen(msg),
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+	};
+
+	ret = nrf_cloud_send(&mqtt_msg);
+	if (ret) {
+		LOG_ERR("Transmission of enqueued device message failed, nrf_cloud_send "
+			"gave error: %d. The message will be re-enqueued and tried again "
+			"later.", ret);
+
+		/* Re-enqueue the message for later retry (creating a new copy of it in heap). */
+		send_device_message(msg);
+
+		/* Increment the failure counter. */
+		send_failure_count += 1;
+
+		/* If we have failed too many times in a row, there is likely a bigger problem,
+		 * and we should reset our connection to nRF Cloud, and wait for a few seconds.
+		 */
+		if (send_failure_count > CONFIG_MAX_CONSECUTIVE_SEND_FAILURES) {
+			/* Disconnect. */
+			disconnect_cloud();
+
+			/* Wait for a few seconds before trying again. */
+			k_sleep(K_SECONDS(CONFIG_CONSECUTIVE_SEND_FAILURE_COOLDOWN_SECONDS));
+		}
+	} else {
+		LOG_DBG("Enqueued device message consumed successfully");
+
+		/* Reset the failure counter, since we succeeded. */
+		send_failure_count = 0;
+	}
+
+	/* Clean up. */
+	k_free(msg);
+
+	return ret;
+}
+
+
+int send_device_message(const char *const msg)
+{
+	/* Copy message data onto the heap (will be freed by the message consumer when done). */
+
+	char *msgbuf = k_malloc(strlen(msg) + 1);
+
+	if (!msgbuf) {
+		LOG_ERR("Could not alloc memory for new device message");
+		return -ENOMEM;
+	}
+	strcpy(msgbuf, msg);
+	LOG_DBG("Enqueued message: %s", log_strdup(msgbuf));
+
+	/* Attempt to append data onto message queue. */
+	if (k_msgq_put(&device_message_queue, &msgbuf, K_NO_WAIT)) {
+		LOG_ERR("Device message rejected, outgoing message queue is full");
+		k_free(msgbuf);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+
+int send_device_message_cJSON(cJSON *msg_obj)
+{
+	int ret = 0;
+	char *msg = NULL;
+
+	if (!msg_obj) {
+		LOG_ERR("Cannot send NULL device message object");
+		return -EINVAL;
+	}
+
+	/* Convert message object to a string. */
+	msg = cJSON_PrintUnformatted(msg_obj);
+	if (msg == NULL) {
+		LOG_ERR("Failed to convert cJSON device message object to string");
+		return -ENOMEM;
+	}
+
+	/* Send the string as a device message. */
+	ret = send_device_message(msg);
+
+	/* Clean up */
+	k_free(msg);
+	return ret;
+}
+
+/**
+ * @brief Reset connection to nRF Cloud, and reset connection status event state.
+ * For internal use only. Externally, disconnect_cloud() may be used instead.
+ */
+static void reset_cloud(void)
+{
+	int err;
+
+	/* Wait for a few seconds to help residual events settle. */
+	LOG_INF("Resetting nRF Cloud transport");
+	k_sleep(K_SECONDS(20));
+
+	/* Reset cloud connection completely. */
+	err = nrf_cloud_disconnect();
+	if (err) {
+		LOG_ERR("Could not disconnect from nRF Cloud, error %d", err);
+	}
+
+	err = nrf_cloud_uninit();
+	if (err) {
+		LOG_ERR("Could not deinitialize nRF Cloud lib, error %d", err);
+	}
+
+	LOG_INF("nRF Cloud transport has been reset");
```

Comment:
I'm on the fence, it certainly could be, but on the flip side, failing to uninit does not necessarily mean that the transport was not reset (it usually just means the transport was **_already_** reset.

I'll go ahead and put it in an else unless the above changes your mind.

Reply:
i didn't actually look at the implementation of `nrf_cloud_uninit` when asking the question 😅 
after looking, if `nrf_cloud_uninit` returns `-EBUSY`, the "transport"/library was definitely not de-initialized.
only in the case of an active FOTA job... so not sure if that matters.

Reply:
Okay, how about this; I'll check for an error code from both cloud_disconnect and cloud_uninit, and if **_either_** of them returns nonzero, I will print "nrf Cloud transport could not be fully reset, continuing anyways"

I think that will be as accurate to the state of things as possible without adding so much bulk that the function becomes unreadable to a newcomer

Reply:
you can remove the disconnect...

```
/**
 * @brief Uninitialize nRF Cloud; disconnects cloud connection
 *  and cleans up allocated memory.
 *
 * @note If nRF Cloud FOTA is enabled and a FOTA job is active
 *  uninit will not be performed.
 *
 * @retval 0      If successful.
 * @retval -EBUSY If a FOTA job is in progress.
 *                Otherwise, a (negative) error code is returned.
 */
int nrf_cloud_uninit(void);
```

Reply:
Alright, I have removed the explicit call to cloud_disconnect, and I have added dedicated messages for the three return value cases of nrf_cloud_uninit

Reply:
I still need to do some disconnection testing to verify that I haven't broken anything, but I am 99% certain what I just pushed will work. Mind taking one last look, @jayteemo?